rat/compiler.py

```diff
@@ -1,488 +1,439 @@
-            # 2. If a variable is marked as primary, then it is the primary variable.
-            # 3. If there is no marked primary variable, then all variables with dataframes are treated as prime.
-            # 4. If there are no variables with dataframes, the leftmost one is the primary one
-            # 5. It is an error if no primary variable can be identified
-            primary_symbol: ast.PrimeableExpr = None
-            # Rule 2
-            for primeable_symbol in ast.search_tree(top_expr, ast.PrimeableExpr):
-                if primeable_symbol.prime:
-                    if primary_symbol is None:
-                        primary_symbol = primeable_symbol
+            msg = (
+                f"No marked primary variable and at least {candidates[0].name} and {candidates[1].name} are"
+                " candidates. A primary variable should be marked manually"
+            )
+            raise CompileError(msg, candidates[0])
+
+    if len(candidates) == 0:
+        msg = f"No primary variable found on line (this means there are no candidate variables)"
+        raise CompileError(msg, statement)
+
+
+# 1. If the variable on the left appears also on the right, mark this
+#   statement to be code-generated with a scan and also make sure that
+#   the right hand side is shifted appropriately
+@dataclass
+class CreateAssignedVariablesWalker(RatWalker):
+    variable_table: VariableTable
+
+    def walk_Statement(self, node: ast.Statement):
+        if node.op == "=":
+            assigned_name = self.walk(node.left)
+            if not assigned_name:
+                msg = "The left hand side of an assignment must be a variable"
+                raise CompileError(msg, node.left)
+            return assigned_name
+
+    def walk_Variable(self, node: ast.Variable):
+        self.variable_table[node.name] = AssignedVariableRecord(node.name, len(node.arglist) if node.arglist else 0)
+        return True
+
+
+@dataclass
+class CreateVariableWalker(RatWalker):
+    variable_table: VariableTable
+    data: Dict[str, pandas.DataFrame]
+    in_control_flow: ContextStack[bool] = field(default_factory=lambda: ContextStack(False))
+
+    def walk_Statement(self, node: ast.Statement):
+        if node.op != "=":
+            self.walk(node.left)
+        self.walk(node.right)
+
+    def walk_IfElse(self, node: ast.IfElse):
+        with self.in_control_flow.push(True):
+            self.walk(node.predicate)
+
+        self.walk(node.left)
+        self.walk(node.right)
+
+    def walk_Variable(self, node: ast.Variable):
+        unallocated = node.name not in self.variable_table
+        subscript_names = _get_subscript_names(node)
+
+        if node.arglist:
+            argument_count = len(node.arglist)
+        else:
+            argument_count = 0
+
+        if unallocated:
+            if self.in_control_flow.peek():
+                record = ConstantVariableRecord(node.name, argument_count)
```

Comment:
Am I right that variables defined from data should be designated as `ConstantVariableRecord`? If so, this logic doesn't catch subscript variables. For example, in the eightschools code(with `eightschools.rat` and `eight_schools.csv` in the integration tests directory), `sigma` will be designated with `SampledVariableRecord`

Reply:
Hmm, looking through ConstantVariableRecord.bind there's a lot of ways this can fail and many of these should be fatal. I think the only error we'd want to actually catch and ignore would be the stuff corresponding to:

```
try:
  data_name = get_dataframe_name_by_column_name(self.name, data)
except KeyError:
  raise KeyError(f"{self.name} not found in the input data")
```

Everything else I think should be fatal cause it corresponds to situations like, we found a matching variable in an input dataframe but the subscripts didn't match, or we had a problem with the subscripts, or there were duplicate rows in the input, etc.

`tests/integration/test_model_noncenter.py` works for me, so if it's failing for you and it's because sigma is trying to be a parameter, I'd change this error logic to only ignore the very specific behavior we want to ignore.

So maybe add a:
```
class VariableNotFoundInData(ValueError):
  pass
```

And then throw and catch only that very specific error (so that anything else will explode and get you more info)

Reply:
```
y[school]' ~ normal(theta[school], sigma[school]);
```
Looks like the problem occurs when this is replaced into:
```
y[school]' ~ normal(theta[school], sigma);
```
Are we rolling with mandating subscripts even for entirely row-based statements?

Reply:
Yeah we're requiring subscripts. I forget where we discussed this, but I think this gets rid of any special-ness of data in a dataframe. Maybe long term we change our tune here (cuz it seems awkward to need to specify sigma if they're literally on the same row in a dataframe). This is simpler though, cause we don't need to worry about how data is arranged in a dataframe to make this work.